panther_lights/src/animation/image_animation.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright 2023 Husarion sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <panther_lights/animation/image_animation.hpp>
+
+#include <algorithm>
+#include <cstdint>
+#include <filesystem>
+#include <regex>
+#include <stdexcept>
+#include <string>
+
+#include <yaml-cpp/yaml.h>
+
+#include <ament_index_cpp/get_package_prefix.hpp>
+#include <ament_index_cpp/get_package_share_directory.hpp>
+#include <boost/gil.hpp>
+#include <boost/gil/extension/io/png.hpp>
+#include <boost/gil/extension/numeric/resample.hpp>
+#include <boost/gil/extension/numeric/sampler.hpp>
+
+namespace panther_lights
+{
+
+void ImageAnimation::Initialize(
+  const YAML::Node & animation_description, const std::size_t num_led,
+  const float controller_frequency)
+{
+  Animation::Initialize(animation_description, num_led, controller_frequency);
+
+  if (!animation_description["image"]) {
+    throw std::runtime_error("No 'image' in animation description");
+  }
+
+  const auto image_path = ParseImagePath(animation_description["image"].as<std::string>());
+  gil::rgba8_image_t base_image;
+  gil::read_and_convert_image(std::string(image_path), base_image, gil::png_tag());
+  image_ = RGBAImageResize(base_image, this->GetNumberOfLeds(), this->GetAnimationLength());
+
+  if (animation_description["color"]) {
+    RGBAImageConvertColor(image_, animation_description["color"].as<std::uint32_t>());
+  }
+}
+
+std::vector<std::uint8_t> ImageAnimation::UpdateFrame()
+{
+  std::vector<std::uint8_t> frame;
+  for (std::size_t i = 0; i < this->GetNumberOfLeds(); i++) {
+    auto pixel = gil::const_view(image_)(i, this->GetAnimationIteration());
+    frame.push_back(pixel[0]);
+    frame.push_back(pixel[1]);
+    frame.push_back(pixel[2]);
+    frame.push_back(pixel[3]);
```

Comment:
```suggestion
    frame.insert(frame.end(), pixel.begin(), pixel.end());
```

Reply:
pixel object does not have `begin()` or `end()`